### ISSUE TITLE
Support array/dictionary indexing

### ISSUE THREAD

**Issue by perlun** [perlang-org/perlang#270]
# Moved to GitLab

Please continue to the new version of this issue here: https://gitlab.perlang.org/perlang/perlang/-/issues/270. The GitHub page you are currently reading will not contain the latest information on this issue.

---

#269 adds some new methods, which shows the need for some new language elements:

```perlang
var env = Libc.environ();
print env.get_Item("PATH");
```

The above _works_, but is anywhere near the elegance we are striving for. Here's the suggested syntax that should be supported:

```perlang
var env = Libc.environ();
print env["PATH"];
```

...in other words, the same as in C#, Ruby, Python and JavaScript (but interestingly enough, **not** in Java; or rather, Java only supports it for built-in arrays (`String[]`, `Object[]` etc), not for other types like `Map<TKey, TValue>`.)

Initially, we can make it work just for `Dictionary<TKey, TValue>`, but in the longer run it would be great to support other indexable types as well (`List<T>` and arrays - the latter would need language support to be able to define arrays though, and the former cannot be instantiated from Perlang since we currently don't support neither the `new` keyword nor using generics. :see_no_evil: However, as noted above, we might still _run into_ arrays or lists being returned from library-level functions, so perhaps we should support such indexing in the language right away. :thinking:).

**Comment by perlun** [comment]
This feels like one of the more critical things we are missing at the moment. Let's aim for including it in 0.3.0; we can always bump it a few more versions if needed.

**Comment by perlun** [comment]
Retained open despite #363 being merged because of the missing features: https://github.com/perlang-org/perlang/pull/363/files#r1104893169

**Comment by perlun** [comment]
What we have now is good enough for getting 0.3.0 out the door. Moving the remaining work to the 0.4.0 iteration.

**Comment by perlun** [comment]
This is not yet working for compiled mode (#406). We should fix that, hopefully in the next 0.6.0 iteration.
